src/decode/stream.rs

```diff
@@ -0,0 +1,497 @@
+/// initialize the stream before processing any data.
+#[derive(Debug)]
+enum State<W>
+where
+    W: Write,
+{
+    /// Stream is initialized but header values have not yet been read.
+    Init(W),
+    /// Header values have been read and the stream is ready to process more data.
+    Run(RunState<W>),
+}
+
+/// Error type used to return the currently owned state along with the error on failure.
+struct StreamStateError<W>
+where
+    W: Write,
+{
+    /// State that was owned while the error occured
+    state: Option<State<W>>,
+    /// IO error that occured in that state
+    error: Error,
+}
+
+impl<W> StreamStateError<W>
+where
+    W: Write,
+{
+    /// Creates a new StreamStateError
+    fn new(state: Option<State<W>>, error: Error) -> Self {
+        Self { state, error }
+    }
+}
+
+/// Structures needed while decoding data.
+struct RunState<W>
+where
+    W: Write,
+{
+    decoder: DecoderState<W, LZCircularBuffer<W>>,
+    range: u32,
+    code: u32,
+}
+
+impl<W> RunState<W>
+where
+    W: Write,
+{
+    fn new(decoder: DecoderState<W, LZCircularBuffer<W>>, range: u32, code: u32) -> Self {
+        Self {
+            decoder,
+            range,
+            code,
+        }
+    }
+}
+
+impl<W> std::fmt::Debug for RunState<W>
+where
+    W: Write,
+{
+    fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
+        fmt.debug_struct("RunState")
+            .field("range", &self.range)
+            .field("code", &self.code)
+            .finish()
+    }
+}
+
+/// Lzma decompressor that can process multiple chunks of data using the
+/// `std::io::Write` interface.
+pub struct Stream<W>
+where
+    W: Write,
+{
+    /// Temporary buffer to hold data while the header is being read.
+    tmp: [u8; MAX_TMP_LEN],
+    /// How many bytes of the temp buffer are in use.
+    tmp_len: usize,
+    /// Whether the stream is initialized and ready to process data.
+    /// An `Option` is used to avoid interior mutability when updating the state.
+    state: Option<State<W>>,
+    /// Options given when a stream is created.
+    options: Options,
+}
+
+impl<W> Stream<W>
+where
+    W: Write,
+{
+    /// Initialize the stream. This will consume the `output` which is the sink
+    /// implementing `std::io::Write` that will receive decompressed bytes.
+    pub fn new(output: W) -> Self {
+        Self::new_with_options(Options::default(), output)
+    }
+
+    /// Initialize the stream with the given `options`. This will consume the
+    /// `output` which is the sink implementing `std::io::Write` that will
+    /// receive decompressed bytes.
+    pub fn new_with_options(options: Options, output: W) -> Self {
+        Self {
+            tmp: [0; MAX_TMP_LEN],
+            tmp_len: 0,
+            state: Some(State::Init(output)),
+            options,
+        }
+    }
+
+    /// Get a reference to the output sink
+    pub fn get_ref(&self) -> Option<&W> {
```

Comment:
The name should be clearer. From the call side it's unclear a reference to what is obtained.

Reply:
I changed it to `get_output()` and `get_output_mut()` to be consistent with `LZBuffer`.